### ISSUE TITLE
Add a hook to transform field config before forms are compiled

### ISSUE THREAD

**Issue by kentcdodds** [formly-js/angular-formly#343]
Would be handy if `formly-form` invoked an overridable function in `formlyConfig` that allowed transforming the field's (adding/removing/modifying) before they're passed to the `formly-field` directive. This would enable things like https://github.com/formly-js/angular-formly-json-schema and https://github.com/formly-js/angular-formly-siren-action to transform a json schema or siren action stuff into field configuration that formly understands. @ckniffen I think you have done some work on this for your own project. Any thoughts?

**Comment by ckniffen** [comment]
Actually this would be helpful for when using the same fields array multiple times.  Right now I have to clone each copy of the fields config when creating formly-forms in an ng-repeat where each iteration uses the exact same fields array.

**Comment by kentcdodds** [comment]
Hmmm... Yeah, I suppose that internally we'll do something like this:

```javascript
const fieldTransform = $scope.options.fieldTransform || formlyConfig.extras.fieldTransform;
if (fieldTransform) {
  $scope.fields = fieldTransform($scope.fields);
}
```

That way you could either copy them yourself or simply modify them to maintain references to the correct field.

The reason that I bring this up is that if `formly-field` actually makes a copy for you, then you may be pointing to the field that you create in your controller, but that ends up not being the field that formly actually uses, which could lead to unexpected behavior.

But I think that kind of an implementation in `formly-form` would be pretty flexible.

**Comment by kentcdodds** [comment]
I have pushed some tests for this feature. I've actually already developed this feature and gotten all the tests passing locally, but I want to give someone the chance to make their first OSS contribution, so here are some instructions (if you have already contributed to open source, please let someone else take this one):

- [ ] Watch [this video](https://www.youtube.com/watch?v=QOchwBm9W-g&index=5&list=PLV5CVI1eNcJi7lVVIuNyRhEuck1Z007BH) to learn what you need to do to get things setup
- [ ] Go to [this line](https://github.com/formly-js/angular-formly/blob/30e87b8db29da8555e3abe51fc4ddb6c5df3b840/src/directives/formly-form.test.js#L595) and change `describe.skip` to `describe.only`
- [ ] Run `$ npm test`
- [ ] Notice that all the tests are failing
- [ ] Update formlyApiCheck.js to allow the `formly-form` `options` to have a `fieldTransform` that's a function (somewhere around [here](https://github.com/formly-js/angular-formly/blob/30e87b8db29da8555e3abe51fc4ddb6c5df3b840/src/providers/formlyApiCheck.js#L153))
- [ ] Add a function call right after [this one](https://github.com/formly-js/angular-formly/blob/30e87b8db29da8555e3abe51fc4ddb6c5df3b840/src/directives/formly-form.js#L104) in `formly-form.js` called `setupFields` and declare it below (probably around [setupOptions](https://github.com/formly-js/angular-formly/blob/30e87b8db29da8555e3abe51fc4ddb6c5df3b840/src/directives/formly-form.js#L130)) where you'll put your logic for initializing the `fields` (which means you could probably move [this](https://github.com/formly-js/angular-formly/blob/30e87b8db29da8555e3abe51fc4ddb6c5df3b840/src/directives/formly-form.js#L106) to that new function).
- [ ] Implement `setupFields` to make the tests pass.
- [ ] Once they're passing, change the `describe.only` to `describe` so all the tests run.
- [ ] Commit your changes to the `src/` directory mentioning issue #343 (notice it runs all the tests... twice)
- [ ] Push your changes to your fork, create a PR, get merged, celebrate 🎉

**Comment by kentcdodds** [comment]
Thanks @stevebluck! This has been officially released in `6.14.0`!
